capsules/src/sht3x.rs

```diff
@@ -0,0 +1,242 @@
+        /// Measurement High Repeatability with Clock Stretch Disabled
+        MEASHIGHREP = 0x2400,
+        /// Measurement Medium Repeatability with Clock Stretch Disabled
+        MEASMEDREP = 0x240B,
+        /// Measurement Low Repeatability with Clock Stretch Disabled
+        MEASLOWREP = 0x2416,
+        /// Read Out of Status Register
+        READSTATUS = 0xF32D,
+        /// Clear Status
+        CLEARSTATUS = 0x3041,
+        /// Soft Reset
+        SOFTRESET = 0x30A2,
+        /// Heater Enable
+        HEATEREN = 0x306D,
+        /// Heater Disable
+        HEATERDIS = 0x3066,
+        /// Status Register Heater Bit
+        REGHEATERBIT = 0x0d,
+    }
+}
+
+#[derive(Clone, Copy, PartialEq)]
+enum State {
+    Idle,
+    Read,
+    ReadData,
+}
+
+fn crc8(data: &[u8]) -> u8 {
+    let polynomial = 0x31;
+    let mut crc = 0xff;
+
+    for x in 0..data.len() {
+        crc ^= data[x as usize] as u8;
+        for _i in 0..8 {
+            if (crc & 0x80) != 0 {
+                crc = crc << 1 ^ polynomial;
+            } else {
+                crc = crc << 1;
+            }
+        }
+    }
+    crc
+}
+
+pub struct SHT3x<'a, A: Alarm<'a>> {
+    i2c: &'a dyn i2c::I2CDevice,
+    humidity_client: OptionalCell<&'a dyn kernel::hil::sensors::HumidityClient>,
+    temperature_client: OptionalCell<&'a dyn kernel::hil::sensors::TemperatureClient>,
+    state: Cell<State>,
+    buffer: TakeCell<'static, [u8]>,
+    read_temp: Cell<bool>,
+    read_hum: Cell<bool>,
+    alarm: &'a A,
+}
+
+impl<'a, A: Alarm<'a>> SHT3x<'a, A> {
+    pub fn new(
+        i2c: &'a dyn i2c::I2CDevice,
+        buffer: &'static mut [u8],
+        alarm: &'a A,
+    ) -> SHT3x<'a, A> {
+        SHT3x {
+            i2c: i2c,
+            humidity_client: OptionalCell::empty(),
+            temperature_client: OptionalCell::empty(),
+            state: Cell::new(State::Idle),
+            buffer: TakeCell::new(buffer),
+            read_temp: Cell::new(false),
+            read_hum: Cell::new(false),
+            alarm: alarm,
+        }
+    }
+
+    fn read_humidity(&self) -> ReturnCode {
+        if self.read_hum.get() == true {
+            ReturnCode::EBUSY
+        } else {
+            if self.state.get() == State::Idle {
+                self.read_hum.set(true);
+                self.read_temp_hum()
+            } else {
+                self.read_hum.set(true);
+                ReturnCode::SUCCESS
+            }
+        }
+    }
+
+    fn read_temperature(&self) -> ReturnCode {
+        if self.read_temp.get() == true {
+            ReturnCode::EBUSY
+        } else {
+            if self.state.get() == State::Idle {
+                self.read_temp.set(true);
+                self.read_temp_hum()
+            } else {
+                self.read_temp.set(true);
+                ReturnCode::SUCCESS
+            }
+        }
+    }
+
+    fn read_temp_hum(&self) -> ReturnCode {
+        self.buffer.take().map_or(ReturnCode::ENOMEM, |buffer| {
```

Comment:
From how I understand the driver, the buffer should only be `None` when an I2C operation is currently running. Wouldn't `EBUSY` be a more appropriate `ReturnCode` then?

If you never expect this function to be called directly (always using `read_temperature` and `read_humidity` instead), and the buffer should therefore never be taken on this function invocation, I think it's okay to panic here (as there's a bug in the driver). This could be done using something like `.map_or_else(|| panic("your message"), |buffer| {...})`.

What option you choose is up to you.